grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1132,106 +1140,438 @@ class StockMovementService {
+    //  parent object above available items that was responsible for answering questions
+    List<AvailableItem> applyAllocationRulesOnAvailableItems(List<AvailableItem> availableItems, Integer quantityRequired) {
+
+        log.info "Apply allocation rules on available items "
+        log.info "Quantity Required = " + quantityRequired
+        log.info "Available items = " + availableItems.size()
+
+        availableItems.each {
+            log.info " - Available item " + it.toJson()
+        }
+
+        // If there's only one pickable bin location
+        if (availableItems.count { it.pickable } == 1) {
+
+            // Scenario 1: All stock in default (no bin)
+            // TODO Could we have more than one available items in a default location?
+            // TODO If there's enough available quantity in the default bin, then allocate
+            //  quantity from the default location
+            boolean isAllStockInDefault = availableItems.every { it.isDefaultLocation }
+            if (isAllStockInDefault) {
+                return availableItems
+            }
+
+            // Scenario 2: All stock in one bin
+            // TODO If there's enough available quantity in one internal location, then
+            //  then allocate stock from that location. Otherwise return an error?
+            boolean isAllStockInSingleBinLocation = availableItems.every { it.isBinLocation }
+            if (isAllStockInSingleBinLocation) {
+                return availableItems
+            }
+        }
+        // If there are more than one internal locations, then we apply the following rules
+        else {
+
+            // Scenario 3a: if lot has stock entries in receiving bins
+            // TODO Are receiving locations configured to be picking locations?
+            // TODO If there's available quantity in default + receiving locations, then
+            //  allocate using FIFO algorithm. Need to check with Manon if this is what is
+            //  expected.
+            boolean isAllStockInReceivingLocations = availableItems.every { it.isReceivingLocation }
+            if (isAllStockInReceivingLocations) {
+                return availableItems
+            }
+            // Scenario 3b: if lot has stock entries in default bin
+            // TODO If there's enough quantity available in default location, then
+            //  allocate from default location.
+
+            // Scenario 3c: if lot has stock entries only in hold bins
+            // TODO If lot has stock entries only in hold bin, then throw an exception.
+
+            // Scenario 4: Any other scenario (stock in “real” bin and virtual bin, stock in multiple real bins)
+            // TODO If there's not enough available quantity in default + receiving
+            //  locations, then throw an exception
+            boolean isAllStockInReceivingOrDefault = availableItems.every { it.isReceivingLocation || it.isDefaultLocation }
```

Comment:
Is there an order of precedence for this scenario which is picked first? Plus for this scenario shouldn't we also look at `stock in “real” bin` looks like you only look at receiving or default